File: iOS/TUIKit/TUICallKit/TUICallKit/UI/Recents/TUICallRecordCallsViewModel.m
//
//  TUICallRecordCallsViewModel.m
//  TUICallKit
//
//  Created by noah on 2023/2/28.
//  Copyright © 2023 Tencent. All rights reserved.
//

#import "TUICallRecordCallsViewModel.h"
#import "TUICallEngineHeader.h"
#import "TUIDefine.h"
#import "TUICore.h"
#import "TUICallRecordCallsCellViewModel.h"
#import "TUICallKit.h"
#import "TUICallKitOfflinePushInfoConfig.h"
#import "TUICallKitConstants.h"

@interface TUICallRecordCallsViewModel()

@property (nonatomic, strong) NSMutableArray<TUICallRecordCallsCellViewModel *> *dataSource;
@property (nonatomic, strong) NSMutableArray<TUICallRecordCallsCellViewModel *> *allDataSource;
@property (nonatomic, strong) NSMutableArray<TUICallRecordCallsCellViewModel *> *missedDataSource;
@property (nonatomic, assign) TUICallRecordCallsType recordCallsType;

@end

@implementation TUICallRecordCallsViewModel

- (instancetype)init {
    self = [super init];
    if (self) {
        _recordCallsType = TUICallRecordCallsTypeAll;
        _recordCallsUIStyle = TUICallKitRecordCallsUIStyleMinimalist;
    }
    return self;
}

- (void)repeatCall:(NSIndexPath *)indexPath {
    TUICallRecordCallsCellViewModel *cellViewModel = self.dataSource[indexPath.row];
    if (!cellViewModel.callRecord) {
        return;
    }
    if (TUICallSceneSingle == cellViewModel.callRecord.scene) {
        [self repeatSingleCall:cellViewModel.callRecord];
    }
}

- (void)repeatSingleCall:(TUICallRecords *)callRecord {
    NSString *userId = [callRecord.inviteList firstObject];
    if (TUICallRoleCalled == callRecord.role) {
        userId = callRecord.inviter;
    }
    if (!userId || userId.length <= 0) {
        return;
    }
    [[TUICallKit createInstance] call:userId callMediaType:callRecord.mediaType];
}

- (void)jumpUserInfoController:(NSIndexPath *)indexPath navigationController:(UINavigationController *)nav {
    TUICallRecordCallsCellViewModel *cellViewModel = self.dataSource[indexPath.row];
    
    if (!cellViewModel.callRecord) {
        return;
    }
    
    NSString *groupId = cellViewModel.callRecord.groupId;
    NSString *userId = cellViewModel.callRecord.inviter;
    
    if (TUICallRoleCall == cellViewModel.callRecord.role) {
        userId = [cellViewModel.callRecord.inviteList firstObject];
    }
    
    if (groupId && groupId.length > 0) {
        NSDictionary *param = @{
            TUICore_TUIGroupObjectFactory_GetGroupInfoVC_GroupID:groupId
        };
        
        if (TUICallKitRecordCallsUIStyleClassic == self.recordCallsUIStyle) {
            [nav pushViewController:TUICore_TUIGroupObjectFactory_GetGroupInfoVC_Classic
                              param:param
                          forResult:nil];
        } else {
            [nav pushViewController:TUICore_TUIGroupObjectFactory_GetGroupInfoVC_Minimalist
                              param:param
                          forResult:nil];
        }
    } else if (userId && userId.length > 0) {
        [self getUserOrFriendProfileVCWithUserID:userId
                                       succBlock:^(UIViewController * _Nonnull viewController) {
            [nav pushViewController:viewController animated:YES];
        } failBlock:^(int code, NSString * _Nonnull desc) {
            [TUITool makeToastError:code msg:desc];
        }];
    }
}

- (void)getUserOrFriendProfileVCWithUserID:(NSString *)userID
                                 succBlock:(void(^)(UIViewController *vc))succ
                                 failBlock:(nullable V2TIMFail)fail {
    NSDictionary *param = @{
        TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod_UserIDKey: userID ? : @"",
        TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod_SuccKey: succ ? : ^(UIViewController *vc){},
        TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod_FailKey: fail ? : ^(int code, NSString * desc){}
    };
    
    if (TUICallKitRecordCallsUIStyleClassic == self.recordCallsUIStyle) {
        [TUICore createObject:TUICore_TUIContactObjectFactory
                          key:TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod
                        param:param];
    } else {
        [TUICore createObject:TUICore_TUIContactObjectFactory_Minimalist
                          key:TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod
                        param:param];
    }
}

- (void)queryRecentCalls {
    TUICallRecentCallsFilter *filter = [[TUICallRecentCallsFilter alloc] init];
    __weak __typeof(self) weakSelf = self;
    [[TUICallEngine createInstance] queryRecentCalls:filter succ:^(NSArray<TUICallRecords *> * _Nonnull callRecord) {
        __strong __typeof(weakSelf) strongSelf = weakSelf;
        NSMutableArray *viewModelList = @[].mutableCopy;
        [callRecord enumerateObjectsUsingBlock:^(TUICallRecords * _Nonnull callRecord, NSUInteger idx, BOOL * _Nonnull stop) {
            TUICallRecordCallsCellViewModel *viewModel = [[TUICallRecordCallsCellViewModel alloc] initWithCallRecord:callRecord];
            if (viewModel) {
                [viewModelList addObject:viewModel];
            }
        }];
        [strongSelf updateDataSourceWith:[viewModelList copy]];
        [strongSelf reloadDataSource];
    } fail:^{
    }];
}

- (void)switchRecordCallsType:(TUICallRecordCallsType)recordCallsType {
    self.recordCallsType = recordCallsType;
    [self reloadDataSource];
}

- (void)deleteRecordCall:(NSIndexPath *)indexPath {
    TUICallRecordCallsCellViewModel *cellViewModel = self.dataSource[indexPath.row];
    
    if (!cellViewModel) {
        return;
    }
    
    [self removeObject:cellViewModel];
    [self reloadDataSource:NO];
    
    if (cellViewModel.callRecord.callId) {
        [[TUICallEngine createInstance] deleteRecordCalls:@[cellViewModel.callRecord.callId] succ:^(NSArray<NSString *> * _Nonnull succList) {
        } fail:^{
        }];
    }
}

- (void)clearAllRecordCalls {
    NSArray *callIdList = [self getCallIdListWith:self.dataSource];
    if (!callIdList || callIdList.count <= 0) {
        return;
    }
    
    __weak __typeof(self) weakSelf = self;
    [[TUICallEngine createInstance] deleteRecordCalls:[callIdList copy] succ:^(NSArray<NSString *> * _Nonnull succList) {
        __strong __typeof(weakSelf) strongSelf = weakSelf;
        [strongSelf.dataSource enumerateObjectsUsingBlock:^(TUICallRecordCallsCellViewModel * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            if ([succList containsObject:obj.callRecord.callId]) {
                [strongSelf removeObject:obj];
            }
        }];
        [strongSelf.dataSource removeAllObjects];
        dispatch_async(dispatch_get_main_queue(), ^{
            [strongSelf reloadDataSource];
        });
    } fail:^{
    }];
}

#pragma mark - private method

- (void)removeObject:(TUICallRecordCallsCellViewModel *)viewModel {
    if (!viewModel) {
        return;
    }
    [self.allDataSource removeObject:viewModel];
    [self.missedDataSource removeObject:viewModel];
}

- (void)removeAllObject {
    [self.dataSource removeAllObjects];
    [self.allDataSource removeAllObjects];
    [self.missedDataSource removeAllObjects];
}

- (void)updateDataSourceWith:(NSArray *)viewModelList {
    if (!viewModelList || viewModelList.count <= 0) {
        return;
    }
    [self removeAllObject];
    self.allDataSource = viewModelList.mutableCopy;
    [viewModelList enumerateObjectsUsingBlock:^(TUICallRecordCallsCellViewModel *viewModel, NSUInteger idx, BOOL * _Nonnull stop) {
        if (viewModel.callRecord && TUICallRecordCallsTypeMissed == viewModel.callRecord.result) {
            [self.missedDataSource addObject:viewModel];
        }
    }];
}

- (void)reloadDataSource {
    [self reloadDataSource:YES];
}

- (void)reloadDataSource:(BOOL)isNeedObserved{
    switch (self.recordCallsType) {
        case TUICallRecordCallsTypeAll:
            if (isNeedObserved){
                self.dataSource = [NSMutableArray arrayWithArray:self.allDataSource];
            } else {
                _dataSource = [NSMutableArray arrayWithArray:self.allDataSource];
            }
            break;
        case TUICallRecordCallsTypeMissed:
            if (isNeedObserved){
                self.dataSource = [NSMutableArray arrayWithArray:self.missedDataSource];
            } else {
                _dataSource = [NSMutableArray arrayWithArray:self.missedDataSource];
            }
            break;
        default:
            break;
    }
}

- (NSArray *)getCallIdListWith:(NSArray<TUICallRecordCallsCellViewModel *> *)cellViewModelArray {
    if (!cellViewModelArray || cellViewModelArray.count <= 0) {
        return @[];
    }
    
    NSMutableArray *callIdList = @[].mutableCopy;
    [cellViewModelArray enumerateObjectsUsingBlock:^(TUICallRecordCallsCellViewModel * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if (obj.callRecord && obj.callRecord.callId) {
            [callIdList addObject:obj.callRecord.callId];
        }
    }];
    
    return [callIdList copy];
}

#pragma mark - getter and setter

- (NSMutableArray *)dataSource {
    if (_dataSource == nil) {
        _dataSource = [[NSMutableArray alloc] init];
    }
    return _dataSource;
}

- (NSMutableArray *)allDataSource {
    if (_allDataSource == nil) {
        _allDataSource = [[NSMutableArray alloc] init];
    }
    return _allDataSource;
}

- (NSMutableArray *)missedDataSource {
    if (_missedDataSource == nil) {
        _missedDataSource = [[NSMutableArray alloc] init];
    }
    return _missedDataSource;
}

@end

File: iOS/TUIKit/TUICallKit/TUICallKit-Swift/ViewModel/Component/Recents/TUICallRecordCallsViewModel.swift
//
//  TUICallRecordCallsViewModel.swift
//  
//
//  Created by vincepzhang on 2023/8/28.
//

import Foundation
import UIKit
import TUICallEngine
import TUICore

/// 通话结果
enum TUICallRecordCallsType: Int {
    case all
    case missed
}

/// UI 风格类型
enum TUICallKitRecordCallsUIStyle: Int {
    case classic // 经典风格
    case minimalist // 简约风格
}

class TUICallRecordCallsViewModel {
    
    var dataSource: Observable<[TUICallRecordCallsCellViewModel]> = Observable(Array())
    var allDataSource: [TUICallRecordCallsCellViewModel] = []
    var missedDataSource: [TUICallRecordCallsCellViewModel] = []
    
    var recordCallsUIStyle: TUICallKitRecordCallsUIStyle = .minimalist
    var recordCallsType: TUICallRecordCallsType = .all
    
    func queryRecentCalls() {
        let filter = TUICallRecentCallsFilter()
        TUICallEngine.createInstance().queryRecentCalls(filter: filter, succ: { [weak self] callRecords in
            guard let self = self else { return }
            var viewModelList: [TUICallRecordCallsCellViewModel] = []
            callRecords.forEach { callRecord in
                let viewModel = TUICallRecordCallsCellViewModel(callRecord)
                viewModelList.append(viewModel)
            }
            self.updateDataSource(viewModelList)
        }, fail: {})
    }
    
    func updateDataSource(_ viewModelList: [TUICallRecordCallsCellViewModel]) {
        if viewModelList.isEmpty {
            return
        }
        cleanAllSource()
        allDataSource = viewModelList
        viewModelList.forEach { viewModel in
            if viewModel.callRecord.result == .missed {
                missedDataSource.append(viewModel)
            }
        }
        reloadDataSource()
    }
    
    func switchRecordCallsType(_ type: TUICallRecordCallsType) {
        recordCallsType = type
        reloadDataSource()
    }
    
    func reloadDataSource() {
        switch recordCallsType {
        case .all:
            dataSource.value = allDataSource
        case .missed:
            dataSource.value = missedDataSource
        }
    }
    
    func cleanAllSource() {
        dataSource.value.removeAll()
        allDataSource.removeAll()
        missedDataSource.removeAll()
    }
    
    func cleanSource(viewModel: TUICallRecordCallsCellViewModel) {
        allDataSource.removeAll() { $0.callRecord.callId == viewModel.callRecord.callId }
        missedDataSource.removeAll() { $0.callRecord.callId == viewModel.callRecord.callId }
    }
    
    func cleanSource(callId: String) {
        allDataSource.removeAll() { $0.callRecord.callId == callId }
        missedDataSource.removeAll() { $0.callRecord.callId == callId }
    }
    
    
    func repeatCall(_ indexPath: IndexPath) {
        let cellViewModel = dataSource.value[indexPath.row]
        
        if cellViewModel.callRecord.scene == .single {
            repeatSingleCall(cellViewModel.callRecord)
        }
    }
    
    func repeatSingleCall(_ callRecord: TUICallRecords) {
        var userId = callRecord.inviteList.first
        if callRecord.role == .called {
            userId = callRecord.inviter
        }
        
        guard let userId = userId as? String else { return }
        
        TUICallKit.createInstance().call(userId: userId, callMediaType: callRecord.mediaType)
    }
    
    func deleteAllRecordCalls() {
        var callIdList: [String] = []
        
        if recordCallsType == .all {
            callIdList = getCallIdList(allDataSource)
        } else if recordCallsType == .missed {
            callIdList = getCallIdList(missedDataSource)
        }
        
        TUICallEngine.createInstance().deleteRecordCalls(callIdList, succ: { [weak self] succList in
            guard let self = self else { return }
            
            succList.forEach { callId in
                self.cleanSource(callId: callId)
            }
            
            self.reloadDataSource()
        }, fail: {})
    }
    
    func getCallIdList(_ cellViewModelArray: [TUICallRecordCallsCellViewModel]) -> [String] {
        var callIdList: [String] = []
        
        if cellViewModelArray.isEmpty {
            return callIdList
        }
        
        cellViewModelArray.forEach { obj in
            callIdList.append(obj.callRecord.callId)
        }
        
        return callIdList
    }
    
    func deleteRecordCall(_ indexPath: IndexPath) {
        if indexPath.row < 0 || indexPath.row >= dataSource.value.count {
            return
        }
        let viewModel = dataSource.value[indexPath.row]
        
        TUICallEngine.createInstance().deleteRecordCalls([viewModel.callRecord.callId], succ: { [weak self] _ in
            guard let self = self else { return }
            self.cleanSource(viewModel: viewModel)
            self.reloadDataSource()
            
        }, fail: {})
    }
    
    func jumpUserInfoController(indexPath: IndexPath, navigationController: UINavigationController) {
        if indexPath.row < 0 || indexPath.row >= dataSource.value.count {
            return
        }
        let cellViewModel = dataSource.value[indexPath.row]
        
        let groupId = cellViewModel.callRecord.groupId
        var userId = cellViewModel.callRecord.inviter
        
        if cellViewModel.callRecord.role == .call {
            guard let firstUserId = cellViewModel.callRecord.inviteList.first as? String else { return }
            userId = firstUserId
        }
        
        if !groupId.isEmpty {
            let param: [String: Any] = [TUICore_TUIGroupObjectFactory_GetGroupInfoVC_GroupID: groupId]
            if TUICallKitRecordCallsUIStyle.classic == recordCallsUIStyle {
                navigationController.push(TUICore_TUIGroupObjectFactory_GetGroupInfoVC_Classic, param: param, forResult: nil)
            } else {
                navigationController.push(TUICore_TUIGroupObjectFactory_GetGroupInfoVC_Minimalist, param: param, forResult: nil)
            }
        } else if !userId.isEmpty {
            getUserOrFriendProfileVCWithUserID(userId: userId) { viewController in
                navigationController.pushViewController(viewController, animated: true)
            } fail: { code, desc in
                TUITool.makeToastError(Int(code), msg: desc)
            }
        }
    }
    
    func getUserOrFriendProfileVCWithUserID(userId: String, succ: ((UIViewController) -> Void)?, fail: ((Int, String) -> Void)?) {
        let param: [String: Any] = [
            TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod_UserIDKey: userId ?? "",
            TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod_SuccKey: succ ?? { _ in },
            TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod_FailKey: fail ?? { _, _ in },
        ]
        
        if TUICallKitRecordCallsUIStyle.classic == self.recordCallsUIStyle {
            TUICore.createObject(TUICore_TUIContactObjectFactory,
                                 key: TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod, param: param)
        } else {
            TUICore.createObject(TUICore_TUIContactObjectFactory_Minimalist,
                                 key: TUICore_TUIContactObjectFactory_GetUserOrFriendProfileVCMethod, param: param)
        }
    }
}
